qiskit/circuit/library/standard_gates/multi_control_su2.py

```diff
@@ -0,0 +1,86 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Multi-controlled SU(2) gate.
+"""
+
+from typing import Union, List
+import numpy as np
+from qiskit.circuit import QuantumCircuit, QuantumRegister, Qubit
+from qiskit.circuit.library.standard_gates.x import MCXVChain
+
+
+def linear_depth_mcv(
+    circuit,
+    x,
+    z,
+    controls: Union[QuantumRegister, List[Qubit]],
+    target: Qubit,
+    ctrl_state: str = None,
+):
+    """
+    Apply multi-controlled SU(2) gate [[z*, x],[-x, z]] with x real and z complex.
+    https://arxiv.org/abs/2302.06377
+
+    Args:
+        circuit (QuantumCircuit): The QuantumCircuit object to apply the diagonal operator on.
+        x (float): real parameter for the single-qubit operators
+        z (complex): complex parameter for the single-qubit operators
+        controls (QuantumRegister or list(Qubit)): The list of control qubits
+        target (Qubit): The target qubit
+        ctrl_state (str): control state of the operator SU(2) operator U
+    """
+
+    alpha_r = np.sqrt((np.sqrt((z.real + 1.0) / 2.0) + 1.0) / 2.0)
+    alpha_i = z.imag / (2.0 * np.sqrt((z.real + 1.0) * (np.sqrt((z.real + 1.0) / 2.0) + 1.0)))
+    alpha = alpha_r + 1.0j * alpha_i
+    beta = x / (2.0 * np.sqrt((z.real + 1.0) * (np.sqrt((z.real + 1.0) / 2.0) + 1.0)))
+
+    s_op = np.array([[alpha, -np.conj(beta)], [beta, np.conj(alpha)]])
+
+    # S gate definition
+    s_gate = QuantumCircuit(1)
+    s_gate.unitary(s_op, 0)
+    s_gate = s_gate.to_gate()
```

Comment:
It might be more direct here to just do `from qiskit.extensions import UnitaryGate` and create the gate directly, `s_gate = UnitaryGate(s_op)`

Reply:
Thanks, @ewinston!

Reply:
> The same strategy should work for `mcrz` gates. However the actual `mcrz` gate applies a multi-controlled phase gate (#9202) and `linear_depth_mcv` only works for SU(2) gates.

@adjs @israelferrazaraujo @ewinston 
Regarding  `mcrx`,  `mcry` and  `mcrz`. The approach in  #9574 (that is the pull request  for #9202) has a better decomposition for the 4 and 5- control version of these rotation gates (and matches the 3-control instance). See [graph](https://www.desmos.com/calculator/ghsaxkkmb3). The #9574 method will also work for 1 and 2 controls. Is this worth including in this pull request?

Reply:
Thank you @ewinston and @AlexisRalli. PR #9574 uses "multiplexers" Theorem 8 from https://arxiv.org/abs/quant-ph/0406176  when the number of controls is less than or equal to 6. This is a good strategy.

Multiplexers are already available in qiskit as `ucry`, `ucrx` and `ucrz`.

Reply:
> > The same strategy should work for `mcrz` gates. However the actual `mcrz` gate applies a multi-controlled phase gate (#9202) and `linear_depth_mcv` only works for SU(2) gates.
> 
> @adjs @israelferrazaraujo @ewinston Regarding `mcrx`, `mcry` and `mcrz`. The approach in #9574 (that is the pull request for #9202) has a better decomposition for the 4 and 5- control version of these rotation gates (and matches the 3-control instance SU2 approach here). See [graph](https://www.desmos.com/calculator/ghsaxkkmb3). The #9574 method will also work for 1 and 2 controls. Is it worth including the method in #9574 for up to 5-controls in this pull request and keep SU(2) approach for 6+ controls?

@AlexisRalli Yes it's worth it. Can you make the change using the ucry and ucrx gates? Or can I do it and include you as a co-author of the commit?

Reply:
I changed it so that the new function is only applied when the number of controls is greater than or equal to 4.